osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/InstallationService.java

```diff
@@ -53,58 +52,34 @@
-    }
-
-    private void reEncryptMasterKey(final SmartMeteringDeviceDto smartMeteringDevice) throws FunctionalException {
-        if (ArrayUtils.isEmpty(smartMeteringDevice.getMasterKey())) {
-            return;
-        }
-        final byte[] reEncryptedMasterKey = this.securityKeyService
-                .reEncryptKey(smartMeteringDevice.getMasterKey(), SecurityKeyType.E_METER_MASTER);
-        smartMeteringDevice.setMasterKey(reEncryptedMasterKey);
-    }
-
-    private void reEncryptAuthenticationKey(final SmartMeteringDeviceDto smartMeteringDevice)
-            throws FunctionalException {
-
-        if (ArrayUtils.isEmpty(smartMeteringDevice.getAuthenticationKey())) {
-            return;
-        }
-        final byte[] reEncryptedAuthenticationKey = this.securityKeyService
-                .reEncryptKey(smartMeteringDevice.getAuthenticationKey(), SecurityKeyType.E_METER_AUTHENTICATION);
-        smartMeteringDevice.setAuthenticationKey(reEncryptedAuthenticationKey);
-    }
-
-    private void reEncryptEncryptionKey(final SmartMeteringDeviceDto smartMeteringDevice) throws FunctionalException {
-
-        if (ArrayUtils.isEmpty(smartMeteringDevice.getGlobalEncryptionUnicastKey())) {
-            return;
-        }
-        final byte[] reEncryptedEncryptionKey = this.securityKeyService
-                .reEncryptKey(smartMeteringDevice.getGlobalEncryptionUnicastKey(), SecurityKeyType.E_METER_ENCRYPTION);
-        smartMeteringDevice.setGlobalEncryptionUnicastKey(reEncryptedEncryptionKey);
-    }
-
-    private void reEncryptMbusDefaultKey(final SmartMeteringDeviceDto smartMeteringDevice) throws FunctionalException {
-
-        if (ArrayUtils.isEmpty(smartMeteringDevice.getMbusDefaultKey())) {
-            return;
+    private void storeNewKeys(final SmartMeteringDeviceDto smartMeteringDeviceDto) {
+        byte[][] securityKeys = new byte[4][];
+
+        try {
+            SecurityKeyType[] securityKeyTypes = { SecurityKeyType.E_METER_MASTER, SecurityKeyType.E_METER_AUTHENTICATION,
+                    SecurityKeyType.E_METER_ENCRYPTION, SecurityKeyType.G_METER_MASTER };
+
+            securityKeys[0] = this.securityKeyService.rsaDecrypt(smartMeteringDeviceDto.getMasterKey(),
+                    SecurityKeyType.E_METER_MASTER);
+            securityKeys[1] = this.securityKeyService.rsaDecrypt(smartMeteringDeviceDto.getAuthenticationKey(),
+                    SecurityKeyType.E_METER_AUTHENTICATION);
+            securityKeys[2] = this.securityKeyService.rsaDecrypt(smartMeteringDeviceDto.getGlobalEncryptionUnicastKey(),
+                    SecurityKeyType.G_METER_ENCRYPTION);
+            securityKeys[3] = this.securityKeyService.rsaDecrypt(smartMeteringDeviceDto.getMbusDefaultKey(),
+                    SecurityKeyType.G_METER_MASTER);
+
+            securityKeyService.storeNewKeys(smartMeteringDeviceDto.getDeviceIdentification(), securityKeyTypes,
+                    securityKeys);
+        } catch (FunctionalException e) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1166)